gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/UnixTimestampRecursiveCopyableDataset.java

```diff
@@ -0,0 +1,170 @@
+    this.currentTime = LocalDateTime.now(this.dateTimeZone);
+  }
+
+  private enum SnapshotSelectionPolicy {
+    EARLIEST, LATEST, ALL
+  }
+
+  class TimestampPathFilter implements PathFilter {
+
+    @Override
+    public boolean accept(Path path) {
+
+      LocalDate endDate = currentTime.toLocalDate();
+      LocalDate startDate = endDate.minus(lookbackPeriod);
+      Matcher matcher = timestampPattern.matcher(path.toString());
+      if (!matcher.matches()) {
+        return false;
+      }
+      Long timestamp = Long.parseLong(matcher.group(1));
+      LocalDate dateOfTimestamp = new LocalDateTime(timestamp, dateTimeZone).toLocalDate();
+      return !(dateOfTimestamp == null || dateOfTimestamp.isAfter(endDate) || dateOfTimestamp.isEqual(startDate)
+          || dateOfTimestamp.isBefore(startDate));
+    }
+  }
+
+  @Override
+  protected List<FileStatus> getFilesAtPath(FileSystem fs, Path path, PathFilter fileFilter)
+      throws IOException {
+
+    // Filter files by lookback period (fileNames >= startDate and fileNames <= endDate)
+    PathFilter andPathFilter = new AndPathFilter(fileFilter, new TimestampPathFilter());
+    List<FileStatus> files = super.getFilesAtPath(fs, path, andPathFilter);
+
+    if (SnapshotSelectionPolicy.ALL == snapshotSelectionPolicy) {
+      return files;
+    }
+
+    Map<String, Map<LocalDate, Long>> pathDateTimestampMap = new HashMap<>();
+
+    // Now select files per day based on snapshot selection policy
+    for (FileStatus fileStatus : files) {
+      String fullPath = fileStatus.getPath().toString();
+      Matcher matcher = timestampPattern.matcher(fullPath);
+      if (!matcher.matches()) {
+        continue;
+      }
+      String timestampStr = matcher.group(1);
+      String pathToSnapshotFolder = fullPath.substring(0, fullPath.indexOf(timestampStr));
+      if (!pathDateTimestampMap.containsKey(pathToSnapshotFolder)) {
+        pathDateTimestampMap.put(pathToSnapshotFolder, new HashMap<>());
+      }
+      Map<LocalDate, Long> dateFileStatusMap = pathDateTimestampMap.get(pathToSnapshotFolder);
+      Long unixTimestamp = Long.parseLong(timestampStr);
+      LocalDate fileDate = new LocalDateTime(unixTimestamp).toLocalDate();
+      if (!dateFileStatusMap.containsKey(fileDate)) {
+        dateFileStatusMap.put(fileDate, unixTimestamp);
+        continue;
+      }
+      Long fromMap = dateFileStatusMap.get(fileDate);
+
+      switch (snapshotSelectionPolicy) {
+        case EARLIEST:
+          if (unixTimestamp < fromMap) {
+            dateFileStatusMap.put(fileDate, unixTimestamp);
+          }
+          break;
+        case LATEST:
+          if (unixTimestamp > fromMap) {
+            dateFileStatusMap.put(fileDate, unixTimestamp);
+          }
+          break;
+        default:
+          throw new RuntimeException("Unsupported snapshot selection policy");
+      }
+    }
+
+    List<FileStatus> result = Lists.newArrayList();
+    for (FileStatus fileStatus : files) {
```

Comment:
Seems like lines 150-157 are repeated from the for loop earlier. Any chance we can avoid the recompute?

Reply:
Can't we group the files based on timestamp in the Hashmap?